File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n, 1);
        int prefix = 1, postfix = 1;
        for(int i = 0; i < n; i++){
            result[i] = prefix;
            prefix *= nums[i];
        }
        for(int i = n - 1; i >= 0; i--){
            result[i] *= postfix;
            postfix *= nums[i];
        }
        return result;
    }
};

// [1, 2, 3, 4]
// [1, 1, 2, 6]
// [24, 12, 8,  6]

// Time: O(N)
// Spce: O(1)
/*
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> left(n + 1), right(n + 1), ans(n);
        left[0] = 1;
        for(int i = 0; i < n; i++){
            left[i + 1] = left[i] * nums[i];
        }
        right[n - 1] = 1;
        for(int i = n - 1; i > 0; i--){
            right[i - 1] = right[i] * nums[i];
        }
        for(int i = 0; i < n; i++){
            ans[i] = left[i] * right[i];
        }
        return ans;
    }
};
*/

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left, right, res = [1] * n, [1] * n, [1] * n
        for i in range(1, n):
            left[i] = left[i - 1] * nums[i - 1]
        for i in range(n - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1]
        for i in range(n):
            res[i] = left[i] * right[i]
        return res
'''
Time complexity: O(n)
Space complexity: O(n)
'''